src/main/java/de/tum/in/www1/artemis/service/LearningPathService.java

```diff
@@ -389,6 +402,320 @@ private void generateNgxGraphRepresentationForRelation(CompetencyRelation relati
+     */
+    private HashMap<Long, Long> getRelationsOfTypeCompetencyMapping(Set<Competency> competencies, HashMap<Long, Set<Long>> matchingClusters,
+            HashMap<Long, Set<Long>> priorCompetencies, CompetencyRelation.RelationType type) {
+        HashMap<Long, Long> map = new HashMap<>();
+        for (var competency : competencies) {
+            if (!map.containsKey(competency.getId())) {
+                long numberOfRelations = competencyRelationRepository.countRelationsOfTypeBetweenCompetencyGroups(matchingClusters.get(competency.getId()), type,
+                        priorCompetencies.get(competency.getId()));
+                // add for each in cluster to reduce database calls (once per cluster)
+                matchingClusters.get(competency.getId()).forEach(id -> map.put(id, numberOfRelations));
+            }
+        }
+        return map;
+    }
+
+    /**
+     * Gets the set of competencies that are themselves not mastered and no matching competency is mastered.
+     *
+     * @param competencies the set of competencies that should be filtered
+     * @param state        the current state of the recommendation system
+     * @return set of pending competencies
+     */
+    private Set<Competency> getPendingCompetencies(Set<Competency> competencies, RecommendationState state) {
+        Set<Competency> pendingCompetencies = new HashSet<>(competencies);
+        pendingCompetencies.removeIf(competency -> state.masteredCompetencies.contains(competency.getId())
+                || state.matchingClusters.get(competency.getId()).stream().anyMatch(state.masteredCompetencies::contains));
+        return pendingCompetencies;
+    }
+
+    /**
+     * Generates a recommended ordering of competencies.
+     *
+     * @param pendingCompetencies the set of pending competencies
+     * @param state               the current state of the recommendation system
+     * @return recommended ordering of competencies
+     */
+    private List<Long> simulateProgression(Set<Competency> pendingCompetencies, RecommendationState state) {
+        List<Long> recommendedOrder = new ArrayList<>();
+        while (!pendingCompetencies.isEmpty()) {
+            HashMap<Long, Double> utilities = computeUtilities(pendingCompetencies, state);
+            var maxEntry = utilities.entrySet().stream().max(Comparator.comparingDouble(Map.Entry::getValue));
+            // is present since outstandingCompetencies is not empty
+            Long competencyId = maxEntry.get().getKey();
+
+            // add competency to recommended order
+            recommendedOrder.add(competencyId);
+
+            // simulate completion of competency
+            state.masteredCompetencies.add(competencyId);
+            pendingCompetencies.removeIf(competency -> state.masteredCompetencies.contains(competency.getId())
+                    || state.matchingClusters.get(competency.getId()).stream().anyMatch(state.masteredCompetencies::contains));
```

Comment:
I need the competency objects later on. Since these are only references the memory usage is comparable to working with ids. I prefer to stay with this implementation to keep the code readable rather than having a map to fetch the competencies again.